src/asgen/backends/alpinelinux/apkindexutils.d

```diff
@@ -0,0 +1,149 @@
+module asgen.backends.alpinelinux.apkindexutils;
+
+import std.algorithm : remove;
+import std.algorithm.iteration : map;
+import std.algorithm.searching : canFind;
+import std.array : join, split;
+import std.file : exists;
+import std.format : format;
+import std.path : buildPath;
+import std.range : empty, InputRange, isForwardRange;
+import std.string : splitLines, startsWith, strip;
+import std.utf : validate;
+
+import std.stdio;
+
+import asgen.backends.alpinelinux.apkpkg;
+import asgen.downloader : Downloader, DownloadException;
+import asgen.logging : logDebug;
+import asgen.utils : isRemote;
+
+/**
+* Struct representing a block inside of an APKINDEX. Each block, seperated by
+* a newline, contains information about exactly one package.
+*/
+struct ApkIndexBlock {
+    string arch;
+    string maintainer;
+    string pkgname;
+    string pkgversion;
+    string pkgdesc;
+
+    @property string archiveName () {
+        return format ("%s-%s.apk", this.pkgname, this.pkgversion);
+    }
+}
+
+/**
+* Range for looping over the contents of an APKINDEX, block by block.
+*/
+struct ApkIndexBlockRange {
+    this (string contents)
+    {
+        this.lines = contents.splitLines;
+        this.getNextBlock();
+    }
+
+    @property ApkIndexBlock front () const {
+        return this.currentBlock;
+    }
+
+    @property bool empty () {
+        return this.m_empty;
+    }
+
+    void popFront ()
+    {
+        this.getNextBlock ();
+    }
+
+    @property ApkIndexBlockRange save () { return this; }
+
+private:
+    void getNextBlock () {
+        string[] completePair;
+        uint iterations = 0;
+        foreach (currentLine; this.lines[this.lineDelta .. $]) {
```

Comment:
This needs to set `currentBlock = ApkIndexBlock();` at the beginning of the function to clean up the current front block for later refilling.